### PR TITLE
STM32G4 Support for PWR, FLASH, RCC, TIMER, DMAMUX, DMA, ADC, OPAMP & USB

### PR DESCRIPTION
**Pull request by ben-brewer-codethink**
Tried to be as in-keeping as possible with other implementations, copyrights sections ended up a bit crowded as I included every contributor from every file I based off.

### CODE REVIEW

**Comment by ben-brewer-codethink** [comment]
I think this is ready for review if possible.

**Comment by karlp** [comment]
sure youv'e got enouigh commits here? :)

File: include/libopencm3/stm32/g4/pwr.h [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+ * @date 29 July 2020
+ *
+ * LGPL License Terms @ref lgpl_license
+ *  */
+
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2016 Benjamin Levine <benjamin@jesco.karoo.co.uk>
+ * Copyright (C) 2020 Ben Brewer <ben.brewer@codethink.co.uk>
```
Comment (karlp) [inline include/libopencm3/stm32/g4/pwr.h:22]: I find it _realllllly_ problematic when people add copyright claims, but can't get their own stories straight.  Is this BenjaminL, Guillaume and BenB, or just Benjamin and Ben? What, if anything, is _actually_ copyright BenjaminL, in 2016, in this file?  based on a file? sure, APIs aren't copyrightable directly, witness google/oracle, so this is.... unclear/misleading, but... whatever, nothing new I guess?

File: lib/stm32/g4/pwr.c [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+{
+	uint32_t reg32;
+
+	reg32 = PWR_CR2;
+	reg32 &= ~(PWR_CR2_PLS_MASK << PWR_CR2_PLS_SHIFT);
+	PWR_CR2 = (reg32 | (pvd_level << PWR_CR2_PLS_SHIFT) | PWR_CR2_PVDE);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Disable Power Voltage Detector.
```
Comment (karlp) [inline lib/stm32/g4/pwr.c:97]: why do we only _sometimes_ have doxygen?

File: include/libopencm3/stm32/g4/rcc.h [hunk 3]
```diff
@@ -7,15 +7,34 @@
...
 /*
  * This file is part of the libopencm3 project.
  *
+ * Copyright (C) 2009 Uwe Hermann <uwe@hermann-uwe.de>
+ * Copyright (C) 2009 Federico Ruiz-Ugalde <memeruiz at gmail dot com>
+ * Copyright (C) 2011 Fergus Noble <fergusnoble@gmail.com>
+ * Copyright (C) 2011 Stephen Caudle <scaudle@doceme.com>
  * Copyright (C) 2020 Karl Palsson <karlp@tweak.net.au>
+ * Copyright (C) 2020 Sam Kirkham <sam.kirkham@codethink.co.uk>
+ * Copyright (C) 2020 Ben Brewer <ben.brewer@codethink.co.uk>
```
Comment (karlp) [inline include/libopencm3/stm32/g4/rcc.h:37]: so run it by me again, I made a new file from scratch, you (ben? sam?) added to it, and have added 4 new authors back to 2009?  I.... struggle to follow how that works.

File: include/libopencm3/stm32/g4/rcc.h [hunk 4]
```diff
@@ -192,27 +234,30 @@
...
 
 /* Division for PLLSAI3CLK, 0 == 7, 1 == 17 */
-#define RCC_PLLCFGR_PLLP		BIT17
+
 #define RCC_PLLCFGR_PLLP_DIV7		0
-#define RCC_PLLCFGR_PLLP_DIV17		RCC_PLLCFGR_PLLP
-#define RCC_PLLPEN			(1 << 16)
+#define RCC_PLLCFGR_PLLP_DIV17		1
+#define RCC_PLLCFGR_PLLP_SHIFT		17
+#define RCC_PLLCFGR_PLLP_MASK		0x1
```
Comment (karlp) [inline include/libopencm3/stm32/g4/rcc.h:244]: we generally don't use mask/shift definitions for single bit fields, is there any particular reason this had to be donne?

File: include/libopencm3/stm32/g4/rcc.h [hunk 5]
```diff
@@ -192,27 +234,30 @@
...
+
+#define RCC_PLLCFGR_PLLPEN		(1 << 16)
 
 /** @defgroup rcc_pllcfgr_plln RCC_PLLCFGR PLLN values
- * Allowed values 8 <= n <= 127, VCO output between 64 and 344 MHz
+ * Allowed values 8 <= n <= 127, VCO output between 96 and 344 MHz
```
Comment (karlp) [inline include/libopencm3/stm32/g4/rcc.h:249]: I see where you got this from, but it's actually _also_ wrong, (in range 2, it's only allowed up to 128) and the refman in rev5 even dropped the numbers and just said, "within the range allowed in the datasheet" :)

File: include/libopencm3/stm32/g4/rcc.h [hunk 6]
```diff
@@ -649,40 +696,53 @@
...
-#define RCC_CRRCR_HSI48VAL_MASK		0x1ff
-#define RCC_CRRCR_HSI48VAL_SHIFT	7
+#define RCC_CRRCR_HSI48CAL_MASK		0x1ff
+#define RCC_CRRCR_HSI48CAL_SHIFT	7
 
-#define RCC_CRRCR_HSI48RDY	BIT1
-#define RCC_CRRCR_HSI48ON	BIT0
+#define RCC_CRRCR_HSI48RDY		(1 << 1)
+#define RCC_CRRCR_HSI48ON		(1 << 0)
 /**@}*/
```
Comment (karlp) [inline include/libopencm3/stm32/g4/rcc.h:731]: As far as I can tell, 99.9% of changes above this point are whitespace.  Please, just don't do that.  At the very least, don't do it in the same commit.

File: include/libopencm3/stm32/g4/rcc.h [hunk 7]
```diff
@@ -691,242 +751,303 @@ extern uint32_t rcc_apb2_frequency;
...
+	uint8_t flash_waitstates;
+	uint32_t ahb_frequency;
+	uint32_t apb1_frequency;
+	uint32_t apb2_frequency;
+};
+
+extern const struct rcc_clock_scale rcc_hsi_configs[RCC_CLOCK_3V3_END];
+extern const struct rcc_clock_scale rcc_hse_8mhz_3v3[RCC_CLOCK_3V3_END];
+extern const struct rcc_clock_scale rcc_hse_12mhz_3v3[RCC_CLOCK_3V3_END];
+extern const struct rcc_clock_scale rcc_hse_16mhz_3v3[RCC_CLOCK_3V3_END];
```
Comment (karlp) [inline include/libopencm3/stm32/g4/rcc.h:784]: This is it, the clock scale struct, that's all that's new in this file, apart from the ~standard function declarations. I don't buy it at all that this gets 4 new authros and a 2009 date when it's just getting definititions that are "compatible" with another file, and I _certainly_ don't think that the massive whitespace changes are useful in this commit, if ever.

**Comment by karlp** [comment]
Running comment tracking changes for my own notes...

- [x] PWR
- [x] FLASH
- [ ] RCC
- [x] TIMERS
- [X] DMAMUX
- [x] adc extsel refactor
- [x] ADC
- [ ] remainder (not yet reviewed)

File: include/libopencm3/stm32/g4/pwr.h (hunk 1 above)
Comment (ben-brewer-codethink) [reply include/libopencm3/stm32/g4/pwr.h:22]: None of the files are mine originally, I copied and pasted the closest relevant file and made the minimal changes required to make it STM32G4 compatible, so I thought it made sense to copy the original copyright as I guess legally we both have copyright claims to the file? Happy to change it if there's another process to follow.

File: lib/stm32/g4/pwr.c (hunk 2 above)
Comment (ben-brewer-codethink) [reply lib/stm32/g4/pwr.c:97]: I based all the new files on existing files so the level of doxygen commenting in the files is based on the variance in the files I copied from. I can go through and try to make them consistent if there's an example somewhere?

File: include/libopencm3/stm32/g4/rcc.h (hunk 3 above)
Comment (ben-brewer-codethink) [reply include/libopencm3/stm32/g4/rcc.h:37]: The new file is based on a direct copy of an existing file, from my limited understanding that doesn't remove copyright, but if that's not the process for adding copyright to the files then I'm happy to change this.

Comment (ben-brewer-codethink) [reply include/libopencm3/stm32/g4/rcc.h:37]: Ah sorry, yeah I included authors from any file I copied directly from too, I guess that's incorrect? Should I just change this to the direct authors of the new content in the file?

File: lib/stm32/g4/pwr.c (hunk 2 above)
Comment (ben-brewer-codethink) [reply lib/stm32/g4/pwr.c:97]: Ah, in this case I added new functions and forgot, sorry will fix this.

File: include/libopencm3/stm32/g4/rcc.h (hunk 5 above)
Comment (ben-brewer-codethink) [reply include/libopencm3/stm32/g4/rcc.h]: This should be fixed now.

File: include/libopencm3/stm32/g4/rcc.h (hunk 6 above)
Comment (ben-brewer-codethink) [reply include/libopencm3/stm32/g4/rcc.h:713]: This should be fixed now.

File: include/libopencm3/stm32/g4/rcc.h (hunk 7 above)
Comment (ben-brewer-codethink) [reply include/libopencm3/stm32/g4/rcc.h:766]: This should be fixed now.

File: include/libopencm3/stm32/g4/rcc.h (hunk 4 above)
Comment (ben-brewer-codethink) [reply include/libopencm3/stm32/g4/rcc.h]: I think this was done to make the code look more consistent, but you're right; there's insufficient reason to change what was already there. I've fixed now.

File: lib/stm32/g4/pwr.c (hunk 2 above)
Comment (ben-brewer-codethink) [reply lib/stm32/g4/pwr.c:97]: Should be fixed now.

File: include/libopencm3/stm32/g4/pwr.h (hunk 1 above)
Comment (karlp) [reply include/libopencm3/stm32/g4/pwr.h:22]: well, you have a different set of authors in the "Copyright" and in the "Authors" and I don't think that sounds right :)

File: lib/stm32/g4/pwr.c [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+ * This library supports the power control system for the
+ * STM32G4 series of ARM Cortex Microcontrollers by ST Microelectronics.
+ *
+ * LGPL License Terms @ref lgpl_license
+ */
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2016 Benjamin Levine <benjamin@jesco.karoo.co.uk>
+ * Copyright (C) 2020 Ben Brewer <ben.brewer@codethink.co.uk>
```
Comment (karlp) [inline lib/stm32/g4/pwr.c:24]: same mismatch in this file

File: include/libopencm3/stm32/g4/rcc.h (hunk 6 above)
Comment (karlp) [reply include/libopencm3/stm32/g4/rcc.h:713]: much easier to read now! thanks!

File: lib/stm32/g4/rcc.c [hunk 9]
```diff
@@ -0,0 +1,762 @@
...
+ *
+ * LGPL License Terms @ref lgpl_license
+ */
+
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2009 Federico Ruiz-Ugalde <memeruiz at gmail dot com>
+ * Copyright (C) 2009 Uwe Hermann <uwe@hermann-uwe.de>
+ * Copyright (C) 2010 Thomas Otto <tommi@viadmin.org>
```
Comment (karlp) [inline lib/stm32/g4/rcc.c:24]: again, you've got totally inconsistent authors here.

File: include/libopencm3/stm32/g4/dmamux.h [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+#define DMAMUX_CxCR_SYNC_ID_RESERVED22       22
+#define DMAMUX_CxCR_SYNC_ID_RESERVED23       23
+#define DMAMUX_CxCR_SYNC_ID_RESERVED24       24
+#define DMAMUX_CxCR_SYNC_ID_RESERVED25       25
+#define DMAMUX_CxCR_SYNC_ID_RESERVED26       26
+#define DMAMUX_CxCR_SYNC_ID_RESERVED27       27
+#define DMAMUX_CxCR_SYNC_ID_RESERVED28       28
+#define DMAMUX_CxCR_SYNC_ID_RESERVED29       29
+#define DMAMUX_CxCR_SYNC_ID_RESERVED30       30
+#define DMAMUX_CxCR_SYNC_ID_RESERVED31       31
```
Comment (karlp) [inline include/libopencm3/stm32/g4/dmamux.h:76]: I don't see any reason for the _RESERVEDxx bits to be defined honestly.

File: include/libopencm3/stm32/g4/dmamux.h [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
+#define DMAMUX_CxCR_DMAREQ_ID_RESERVED118      118
+#define DMAMUX_CxCR_DMAREQ_ID_RESERVED119      119
+#define DMAMUX_CxCR_DMAREQ_ID_RESERVED120      120
+#define DMAMUX_CxCR_DMAREQ_ID_RESERVED121      121
+#define DMAMUX_CxCR_DMAREQ_ID_RESERVED122      122
+#define DMAMUX_CxCR_DMAREQ_ID_RESERVED123      123
+#define DMAMUX_CxCR_DMAREQ_ID_RESERVED124      124
+#define DMAMUX_CxCR_DMAREQ_ID_RESERVED125      125
+#define DMAMUX_CxCR_DMAREQ_ID_RESERVED126      126
+#define DMAMUX_CxCR_DMAREQ_ID_RESERVED127      127
```
Comment (karlp) [inline include/libopencm3/stm32/g4/dmamux.h:207]: nor here

File: include/libopencm3/stm32/g4/dmamux.h [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+#define DMAMUX_CxCR_SYNC_ID_RESERVED22       22
+#define DMAMUX_CxCR_SYNC_ID_RESERVED23       23
+#define DMAMUX_CxCR_SYNC_ID_RESERVED24       24
+#define DMAMUX_CxCR_SYNC_ID_RESERVED25       25
+#define DMAMUX_CxCR_SYNC_ID_RESERVED26       26
+#define DMAMUX_CxCR_SYNC_ID_RESERVED27       27
+#define DMAMUX_CxCR_SYNC_ID_RESERVED28       28
+#define DMAMUX_CxCR_SYNC_ID_RESERVED29       29
+#define DMAMUX_CxCR_SYNC_ID_RESERVED30       30
+#define DMAMUX_CxCR_SYNC_ID_RESERVED31       31
```
Comment (karlp) [inline include/libopencm3/stm32/g4/dmamux.h:246]: or here

File: lib/stm32/g4/Makefile [hunk 13]
```diff
@@ -45,6 +45,10 @@ OBJS += pwr.o
 OBJS += rcc.o rcc_common_all.o
 OBJS += timer_common_all.o timer_common_f0234.o
 
+OBJS += usb.o usb_control.o usb_standard.o usb_msc.o
+OBJS += usb_hid.o
```
Comment (karlp) [inline lib/stm32/g4/Makefile:51]: looks fine, but someone remind me to add teh other classes there, see https://github.com/libopencm3/libopencm3/pull/1221 to add after we're all merged in....

Comment (ben-brewer-codethink) [reply lib/stm32/g4/Makefile:51]: > For RCC: what's the significance of 24/48/96/170? Do we really need all of them? are any of them the max for any particular family/variant? On F1/f4, we actually had different maxes, or max to still use usb, so some of these made sense, but some of these seem like "just a random assortment" ? I'd like to only include useful configs out of the box,

It's a bit arbitrary, but I selected like so:
24MHz is in the low-power mode, supporting USB using PLL (26MHz is maximum)
48MHz is the lowest flash wait-state and supports USB using PLL.
96MHz is the highest that supports USB using PLL.
170MHz is the highest the chip supports, CRS is needed for USB.

I guess it's a bit arbitrary, and you can probably tell that we used that sequence before we had USB CRS working

**Comment by ben-brewer-codethink** [comment]
> For RCC: what's the significance of 24/48/96/170? Do we really need all of them? are any of them the max for any particular family/variant? On F1/f4, we actually had different maxes, or max to still use usb, so some of these made sense, but some of these seem like "just a random assortment" ? I'd like to only include useful configs out of the box,

It's a bit arbitrary, but I selected like so:
24MHz is in the low-power mode, supporting USB using PLL (26MHz is maximum)
48MHz is the lowest flash wait-state and supports USB using PLL.
96MHz is the highest that supports USB using PLL.
170MHz is the highest the chip supports, CRS is needed for USB.

I guess it's a bit arbitrary, and you can probably tell that we used that sequence before we had USB CRS working

**Comment by karlp** [comment]
Ok, this has merged all the way to master, _MINUS_ the opamp and dac commits, as we want to try and unify them with existing code more.  I'm going to close this one, but please do re-open new ones for the dac/opamp (separately or together)

**Pull request closed without merging**